### PR TITLE
Bitnami Httpd image

### PR DESCRIPTION
**Pull request by AnmolVirdi**
Things that are left:
1. Generating readme using jinja scripts.
2. Publishing the image on RapidFort
3. Running functional_tests just in case

### CODE REVIEW

File: community_images/common/helpers.sh [hunk 1]
```diff
@@ -285,16 +285,6 @@ function report_pulls()
 }
 
 function finish {
-    if [[ -z "$NAMESPACE_TO_CLEANUP" ]]; then
```
Comment (dreamer-nitj) [inline community_images/common/helpers.sh:288]: why are we removing this?

File: community_images/common/latest_tag.py [hunk 2]
```diff
@@ -17,6 +17,7 @@ def fetch_tags(image_repo):
         image_repo=image_repo[2:]
 
     resp = requests.get(f"https://registry.hub.docker.com/v1/repositories/{image_repo}/tags")
+    tags = resp;
```
Comment (dreamer-nitj) [inline community_images/common/latest_tag.py:20]: why is this required?

File: community_images/httpd/bitnami/functional_test.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+# shellcheck disable=SC1091
+. "${SCRIPTPATH}"/../../common/helpers.sh
+
+HELM_RELEASE=apache-release
+NAMESPACE=$(get_namespace_string "${HELM_RELEASE}")
+REPOSITORY=apache
+IMAGE_REPOSITORY=rapidfort/"$REPOSITORY"
+
+k8s_test()
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/functional_test.sh:16]: we should also add some tests to k8s_test as well. probably the coverage script can be added here as well to begin with.

File: community_images/httpd/bitnami/functional_test.sh [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    # kill docker-compose setup container
+    docker-compose -f "${SCRIPTPATH}"/docker-compose.yml -p "${NAMESPACE}" down
+
+    # clean up docker file
+    rm -rf "${SCRIPTPATH}"/docker-compose.yml
+}
+
+main()
+{
+    # k8s_test
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/functional_test.sh:92]: uncomment them

File: community_images/httpd/bitnami/module_gen.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import argparse
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/module_gen.py:1]: where is this script getting used?

File: community_images/common/latest_tag.py (hunk 2 above)
Comment (AnmolVirdi) [reply community_images/common/latest_tag.py:20]: if resp.status_code <200, the functional would return empty list.

File: community_images/httpd/bitnami/run.sh [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    # fetch service url and store the urls in URLS file
+    rm -f URLS
+    minikube service "${HELM_RELEASE}" -n "${NAMESPACE}" --url >> URLS
+
+    # create ssl certs
+    cleanup_certs
+    create_certs
+
+    # Changing "http" to "https" in the urls file
+    sed -i '2,2s/http/https/' URLS
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/run.sh:55]: why changing the http to https only on a specific line?

File: community_images/httpd/bitnami/run.sh [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    done <URLS
+
+    #Removing urls file
+    rm URLS
+
+    # fetch minikube ip
+    MINIKUBE_IP=$(minikube ip)
+
+    # curl to https url
+    curl http://"${MINIKUBE_IP}" -k
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/run.sh:71]: do you want to make it https instead?

File: community_images/httpd/bitnami/module_gen.py (hunk 5 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/module_gen.py:1]: module_gen.py was created to create a list of all the modules and add them to a file. Further, we could use the generated file by directly copying it's content tot he httpd.conf file. But, python didn't work after the stubbing was initiated. Therefore, I kept the script there for future use.

File: community_images/httpd/bitnami/run.sh [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    # install docker container
+    docker-compose -f "${SCRIPTPATH}"/docker-compose.yml -p "${NAMESPACE}" up -d
+    report_pulls "${IMAGE_REPOSITORY}"
+
+    # sleep for 30 sec
+    sleep 30
+
+    # exec into container and run coverage script
+    pwd
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/run.sh:96]: probably not need

File: community_images/httpd/bitnami/run.sh [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    chmod 777 "${SCRIPTPATH}"/coverage_script.sh
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/run.sh:97]: The permissions can be changed outside once in the repo and this line can be removed

File: community_images/httpd/bitnami/run.sh [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    NON_TLS_PORT=$(docker inspect "${NAMESPACE}"-apache-1 | jq -r ".[].NetworkSettings.Ports.\"8080/tcp\"[0].HostPort")
+    TLS_PORT=$(docker inspect "${NAMESPACE}"-apache-1 | jq -r ".[].NetworkSettings.Ports.\"8443/tcp\"[0].HostPort")
+
+    echo "${NON_TLS_PORT}"
+    echo "${TLS_PORT}"
+
+    # run curl in loop for different endpoints
+    for i in {1..20};
+    do
+        echo "$i"
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/run.sh:113]: echo "attempt $i"

File: community_images/httpd/bitnami/run.sh (hunk 6 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/run.sh:55]: It's because the file will contain only two lines. To successfully curl them, the second url needed to be converted to 'https'

File: t.sh [hunk 11]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (dreamer-nitj) [inline t.sh:1]: is this file needed?

File: community_images/httpd/bitnami/run.sh (hunk 7 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/run.sh:71]: This minikube ip was the first url i guess. It ran successfully with both http and https.

File: community_images/httpd/bitnami/run.sh (hunk 8 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/run.sh:96]: Noted.

File: community_images/httpd/bitnami/functional_test.sh [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    report_pulls "${IMAGE_REPOSITORY}"
+
+    # sleep for few seconds
+    sleep 30
+
+    # get docker host ip
+    HTTPD_HOST=$(docker inspect "${NAMESPACE}" | jq -r ".[].NetworkSettings.Networks[\"${NAMESPACE}\"].IPAddress")
+
+    # testing using apache benchmark tool
+    ab -t 100 -n 10000 -c 10 HTTPD_HOST
```
Comment (dreamer-nitj) [inline community_images/httpd/bitnami/functional_test.sh:59]: we may need to also change the setup.sh file to include this tool since it may not be present by default

File: community_images/httpd/bitnami/run.sh (hunk 9 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/run.sh:97]: Agreed, will do that.

**Review by dreamer-nitj — CHANGES_REQUESTED** [review]
Few comments

File: t.sh (hunk 11 above)
Comment (AnmolVirdi) [reply t.sh:1]: My bad. I'll remove it asap.

File: community_images/common/dockertags.sh [hunk 13]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+# This scripts fetchs all dockertags for a given repo
+
```
Comment (codervinod) [inline community_images/common/dockertags.sh:3]: why is this script changes/added?

File: community_images/common/helpers.sh (hunk 1 above)
Comment (AnmolVirdi) [reply community_images/common/helpers.sh:288]: I guess there was some issue with this part before. Checked it again, it worked without removing this part. Added it again. Thanks.

File: community_images/httpd/bitnami/module_gen.py (hunk 5 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/module_gen.py:1]: Removed module_gen.py since it won't be used in this image. Keeping it safe for future use.

File: community_images/httpd/bitnami/functional_test.sh (hunk 3 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/functional_test.sh:16]: Noted.

File: community_images/common/helpers.sh [hunk 14]
```diff
@@ -4,7 +4,7 @@ set -x
 set -e
 
 DOCKERHUB_REGISTRY="${DOCKERHUB_REGISTRY:-docker.io}"
-RAPIDFORT_ACCOUNT="${RAPIDFORT_ACCOUNT:-rapidfort}"
```
[the remark this answers is not in the document]
Comment (AnmolVirdi) [reply community_images/common/helpers.sh]: Added it back.

File: community_images/common/helpers.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnmolVirdi) [reply community_images/common/helpers.sh]: Added it back.

File: community_images/common/latest_tag.py (hunk 2 above)
Comment (dreamer-nitj) [reply community_images/common/latest_tag.py:20]: can you pls check it once again

File: community_images/httpd/bitnami/run.sh (hunk 10 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/run.sh]: Done

File: community_images/common/dockertags.sh (hunk 13 above)
Comment (AnmolVirdi) [reply community_images/common/dockertags.sh:3]: Seems like it existed in the main branch when I created a branch. Removing it.

File: community_images/httpd/bitnami/functional_test.sh (hunk 12 above)
Comment (AnmolVirdi) [reply community_images/httpd/bitnami/functional_test.sh:59]: Added the changes in setup.sh

**Comment by dreamer-nitj** [comment]
needs_ci

File: scripts/setup.sh [hunk 15]
```diff
@@ -63,6 +63,9 @@ DC_DESTINATION=/usr/local/bin/docker-compose
 sudo curl -L https://github.com/docker/compose/releases/download/${DC_VERSION}/docker-compose-$(uname -s)-$(uname -m) -o $DC_DESTINATION
 sudo chmod 755 $DC_DESTINATION
 
+# Install Apache benchmark testing tool
```
Comment (codervinod) [inline scripts/setup.sh:66]: why are we installing tools for all the containers? This should be part of coverage script where we plan to use these tools.

Comment (AnmolVirdi) [reply scripts/setup.sh:66]: Adding it to the docker_test() of the functional_tests file instead. Works?

**Review by codervinod — APPROVED** [review]
